File: packages/sdk-python/sentinel/wrappers/openai.py
"""OpenAI client instrumentation (sync + async, streaming + non-streaming).

Wraps ``chat.completions.create`` and ``embeddings.create`` on an OpenAI client
instance. Streaming responses are wrapped so token deltas are accumulated and
the span is closed only when the stream is fully consumed.
"""

from __future__ import annotations

import logging
from typing import Any

from ..tracer import Tracer
from ..types import SpanKind, TokenUsage
from ._common import patch_method, safe_getattr, to_serializable

logger = logging.getLogger("sentinel")


def _extract_usage(response: Any) -> TokenUsage | None:
    usage = safe_getattr(response, "usage")
    if usage is None:
        return None
    return TokenUsage(
        prompt_tokens=safe_getattr(usage, "prompt_tokens", default=0) or 0,
        completion_tokens=safe_getattr(usage, "completion_tokens", default=0) or 0,
        total_tokens=safe_getattr(usage, "total_tokens", default=0) or 0,
    )


def _stamp_chat_request(span, kwargs: dict) -> None:
    span.model = kwargs.get("model")
    span.provider = "openai"
    span.set_attribute("stream", bool(kwargs.get("stream")))
    for key in ("temperature", "top_p", "max_tokens", "tools", "tool_choice"):
        if key in kwargs and kwargs[key] is not None:
            span.set_attribute(key, to_serializable(kwargs[key]))
    span.input = to_serializable(kwargs.get("messages"))


class _StreamAccumulator:
    """Collects streamed chat chunks into a final text + tool-call summary."""

    def __init__(self) -> None:
        self.content_parts: list[str] = []
        self.usage: TokenUsage | None = None
        self.finish_reason: str | None = None
        self.chunks = 0

    def add(self, chunk: Any) -> None:
        self.chunks += 1
        choices = safe_getattr(chunk, "choices", default=[]) or []
        for choice in choices:
            delta = safe_getattr(choice, "delta")
            if delta is not None:
                piece = safe_getattr(delta, "content")
                if piece:
                    self.content_parts.append(piece)
            reason = safe_getattr(choice, "finish_reason")
            if reason:
                self.finish_reason = reason
        # Newer OpenAI streams include usage on the final chunk.
        usage = _extract_usage(chunk)
        if usage is not None:
            self.usage = usage

    def finalize(self, span, tracer: Tracer) -> None:
        span.set_attribute("stream_chunks", self.chunks)
        if self.finish_reason:
            span.set_attribute("finish_reason", self.finish_reason)
        if self.usage:
            span.usage = self.usage
        tracer.set_output(span, "".join(self.content_parts))


def _instrument_chat_create(tracer: Tracer, original):
    is_async = _looks_async(original)

    if is_async:

        async def async_create(*args: Any, **kwargs: Any):
            span = tracer.start_span("openai.chat.completions", SpanKind.LLM)
            _stamp_chat_request(span, kwargs)
            try:
                result = await original(*args, **kwargs)
            except BaseException as exc:
                span.record_exception(exc)
                tracer.end_span(span)
                raise
            if kwargs.get("stream"):
                return _wrap_async_stream(result, span, tracer)
            _finalize_chat(result, span, tracer)
            tracer.end_span(span)
            return result

        return async_create

    def sync_create(*args: Any, **kwargs: Any):
        span = tracer.start_span("openai.chat.completions", SpanKind.LLM)
        _stamp_chat_request(span, kwargs)
        try:
            result = original(*args, **kwargs)
        except BaseException as exc:
            span.record_exception(exc)
            tracer.end_span(span)
            raise
        if kwargs.get("stream"):
            return _wrap_sync_stream(result, span, tracer)
        _finalize_chat(result, span, tracer)
        tracer.end_span(span)
        return result

    return sync_create


def _finalize_chat(response: Any, span, tracer: Tracer) -> None:
    span.usage = _extract_usage(response)
    choices = safe_getattr(response, "choices", default=[]) or []
    if choices:
        message = safe_getattr(choices[0], "message")
        span.set_attribute(
            "finish_reason", safe_getattr(choices[0], "finish_reason")
        )
        tracer.set_output(span, to_serializable(message))


def _wrap_sync_stream(stream: Any, span, tracer: Tracer):
    acc = _StreamAccumulator()

    def generator():
        try:
            for chunk in stream:
                acc.add(chunk)
                yield chunk
        except BaseException as exc:
            span.record_exception(exc)
            raise
        finally:
            # Finalization must never prevent the span from being closed and
            # exported, so guard the accumulator step separately.
            try:
                acc.finalize(span, tracer)
            except Exception:  # pragma: no cover - defensive
                logger.exception("sentinel: stream finalize failed")
            tracer.end_span(span)

    return generator()


def _wrap_async_stream(stream: Any, span, tracer: Tracer):
    acc = _StreamAccumulator()

    async def generator():
        try:
            async for chunk in stream:
                acc.add(chunk)
                yield chunk
        except BaseException as exc:
            span.record_exception(exc)
            raise
        finally:
            try:
                acc.finalize(span, tracer)
            except Exception:  # pragma: no cover - defensive
                logger.exception("sentinel: stream finalize failed")
            tracer.end_span(span)

    return generator()


def _instrument_embeddings(tracer: Tracer, original):
    is_async = _looks_async(original)

    def _stamp(span, kwargs):
        span.model = kwargs.get("model")
        span.provider = "openai"
        span.input = to_serializable(kwargs.get("input"))

    if is_async:

        async def async_create(*args: Any, **kwargs: Any):
            with tracer.span("openai.embeddings", SpanKind.EMBEDDING) as span:
                _stamp(span, kwargs)
                result = await original(*args, **kwargs)
                span.usage = _extract_usage(result)
                span.set_attribute(
                    "vector_count", len(safe_getattr(result, "data", default=[]) or [])
                )
                return result

        return async_create

    def sync_create(*args: Any, **kwargs: Any):
        with tracer.span("openai.embeddings", SpanKind.EMBEDDING) as span:
            _stamp(span, kwargs)
            result = original(*args, **kwargs)
            span.usage = _extract_usage(result)
            span.set_attribute(
                "vector_count", len(safe_getattr(result, "data", default=[]) or [])
            )
            return result

    return sync_create


def _looks_async(fn: Any) -> bool:
    import inspect

    return inspect.iscoroutinefunction(fn) or inspect.iscoroutinefunction(
        getattr(fn, "__wrapped__", None)
    )


def instrument(client: Any, tracer: Tracer) -> Any:
    """Patch a sync or async OpenAI client in place."""
    patched_chat = patch_method(
        client,
        "chat.completions.create",
        lambda original: _instrument_chat_create(tracer, original),
    )
    patch_method(
        client,
        "embeddings.create",
        lambda original: _instrument_embeddings(tracer, original),
    )
    if not patched_chat:
        logger.warning(
            "sentinel: could not find chat.completions.create on OpenAI client; "
            "is this a supported client version?"
        )
    return client

File: packages/sdk-typescript/src/wrappers/openai.ts
// OpenAI client instrumentation (non-streaming + streaming).
// Patches `chat.completions.create` on the given client instance.

import type { Tracer } from "../tracer.js";
import type { Span, TokenUsage } from "../types.js";

/* eslint-disable @typescript-eslint/no-explicit-any */
type AnyObj = Record<string, any>;

function extractUsage(resp: AnyObj | undefined): TokenUsage | null {
  const u = resp?.usage;
  if (!u) return null;
  return {
    prompt_tokens: u.prompt_tokens ?? 0,
    completion_tokens: u.completion_tokens ?? 0,
    total_tokens: u.total_tokens ?? (u.prompt_tokens ?? 0) + (u.completion_tokens ?? 0),
  };
}

function stampRequest(span: Span, args: AnyObj): void {
  span.model = args?.model ?? null;
  span.provider = "openai";
  span.setAttribute("stream", Boolean(args?.stream));
  span.input = args?.messages ?? null;
}

function isAsyncIterable(x: unknown): x is AsyncIterable<AnyObj> {
  return (
    x != null &&
    typeof (x as { [Symbol.asyncIterator]?: unknown })[Symbol.asyncIterator] === "function"
  );
}

/** Instrument an OpenAI client (the official `openai` npm package) in place. */
export function instrumentOpenAI<T extends AnyObj>(client: T, tracer: Tracer): T {
  const completions = client?.chat?.completions;
  const original = completions?.create;
  if (typeof original !== "function" || (original as AnyObj).__sentinelWrapped) {
    return client;
  }

  const wrapped = async function (this: unknown, args: AnyObj, ...rest: unknown[]) {
    const span = tracer.startSpan("openai.chat.completions", "llm");
    stampRequest(span, args);

    let result: unknown;
    try {
      result = await original.call(completions, args, ...rest);
    } catch (err) {
      span.recordException(err);
      tracer.endSpan(span);
      throw err;
    }

    if (args?.stream && isAsyncIterable(result)) {
      return wrapStream(result, span, tracer);
    }

    const resp = result as AnyObj;
    span.usage = extractUsage(resp);
    const choice = resp?.choices?.[0];
    if (choice) {
      span.setAttribute("finish_reason", choice.finish_reason ?? null);
      tracer.setOutput(span, choice.message ?? null);
    }
    tracer.endSpan(span);
    return result;
  };

  (wrapped as AnyObj).__sentinelWrapped = true;
  completions.create = wrapped;
  return client;
}

async function* wrapStream(
  stream: AsyncIterable<AnyObj>,
  span: Span,
  tracer: Tracer,
): AsyncGenerator<AnyObj> {
  const parts: string[] = [];
  let usage: TokenUsage | null = null;
  let finish: string | null = null;
  let count = 0;
  try {
    for await (const chunk of stream) {
      count++;
      const choice = chunk?.choices?.[0];
      const piece = choice?.delta?.content;
      if (piece) parts.push(piece);
      if (choice?.finish_reason) finish = choice.finish_reason;
      const u = extractUsage(chunk);
      if (u) usage = u;
      yield chunk;
    }
  } catch (err) {
    span.recordException(err);
    throw err;
  } finally {
    span.setAttribute("stream_chunks", count);
    if (finish) span.setAttribute("finish_reason", finish);
    if (usage) span.usage = usage;
    tracer.setOutput(span, parts.join(""));
    tracer.endSpan(span);
  }
}
